## Лабораторная № 2. Векторное представление текстов

### 1. Скачайте данные отзывов о банковских услугах из папки курса (“data/financial_review.csv”).

In [ ]:
import pandas as pd
pd.options.mode.chained_assignment = None
df = pd.read_csv('vtb_review.csv')
df

Unnamed: 0          bank_href             login  \
0           1051  /bank-moskvy.html               NaN   
1           1052  /bank-moskvy.html               NaN   
2           1053  /bank-moskvy.html               NaN   
3           1054  /bank-moskvy.html               NaN   
4           1055  /bank-moskvy.html               NaN   
...          ...                ...               ...   
2598       16850         /bank/vtb/            Юрий,    
2599       16851         /bank/vtb/           Ольга,    
2600       16852         /bank/vtb/  Старшова Елена,    
2601       16853         /bank/vtb/         Наталья,    
2602       16854         /bank/vtb/   Вернер Сергей,    

                                                 review bank_name  \
0     Сменила пароль в интернет-банке - по-новому не...       втб   
1     Вeб-чат: 17:19:23Спасибо за проявленный интере...       втб   
2     Воруют деньги с карты, давно заметил, что проп...       втб   
3     В "рекламе" и договоре написано, что на социал...       втб   
4     Суть проблемы: открытая на мое имя карта Банка...       втб   
...                                                 ...       ...   
2598  Санкт–Петербург. Отделение №15. ул. Бутлерова ...       втб   
2599  неправомерное подключение доп.услуг06.08.2016 ...       втб   
2600  Добрый день! Оформлен кредит в Банке ВТБ24 28....       втб   
2601  Самый неудобный интернет банк. Нужно прийти в ...       втб   
2602  Решил взять в ВТБ24 кредит под 18.5%, на 3 год...       втб   

                       site          date                  message_href score  \
0     http://bankireview.ru           NaN                           NaN   NaN   
1     http://bankireview.ru           NaN                           NaN   NaN   
2     http://bankireview.ru           NaN                           NaN   NaN   
3     http://bankireview.ru           NaN                           NaN   NaN   
4     http://bankireview.ru           NaN                           NaN   NaN   
...                     ...           ...                           ...   ...   
2598  https://www.sravni.ru  19 окт, 2015  /bank/vtb/otzyv/178721/?page     1   
2599  https://www.sravni.ru  06 дек, 2016  /bank/vtb/otzyv/219166/?page     1   
2600  https://www.sravni.ru  23 авг, 2015  /bank/vtb/otzyv/170635/?page     1   
2601  https://www.sravni.ru  23 авг, 2015  /bank/vtb/otzyv/170659/?page     2   
2602  https://www.sravni.ru  23 авг, 2015  /bank/vtb/otzyv/170668/?page     1   

     status user_href                                     cleaned_review  
0       NaN       NaN  ['сменить', 'пароль', 'интернет', 'банк', 'нов...  
1       NaN       NaN  ['вeб', 'чат', '17', '19', '23спасибо', 'прояв...  
2       NaN       NaN  ['воровать', 'деньга', 'карта', 'давно', 'заме...  
3       NaN       NaN  ['реклама', 'договор', 'написать', 'социальный...  
4       NaN       NaN  ['проблема', 'открытый', 'имя', 'карта', 'банк...  
...     ...       ...                                                ...  
2598    NaN       NaN  ['санкт', 'петербург', 'отделение', '15', 'ул'...  
2599    NaN       NaN  ['неправомерный', 'подключение', 'доп', 'услуг...  
2600    NaN       NaN  ['добрый', 'день', 'оформить', 'кредит', 'банк...  
2601    NaN       NaN  ['самый', 'неудобный', 'интернет', 'банк', 'ну...  
2602    NaN       NaN  ['решить', 'взять', 'втб24', 'кредит', '18', '...  

[2603 rows x 12 columns]

In [ ]:
df.isnull().sum()

Unnamed: 0           0
bank_href            0
login              409
review               0
bank_name            0
site                 0
date               409
message_href      1751
score              409
status            1261
user_href         1261
cleaned_review       0
dtype: int64

In [ ]:
df.dropna(subset=['score'], inplace=True)

In [ ]:
df['score'].isnull().sum()

0

In [ ]:
df['cleaned_review']

409     ['добрый', 'день', 'подавать', 'заявка', 'полу...
410     ['здравствуйте', 'онлайн', 'банк', 'поступить'...
411     ['обращаться', 'данный', 'отделение', 'оформле...
412     ['добрый', 'день', 'хотеть', 'поделиться', 'св...
413     ['выражать', 'благодарность', 'надежда', 'пань...
                              ...                        
2598    ['санкт', 'петербург', 'отделение', '15', 'ул'...
2599    ['неправомерный', 'подключение', 'доп', 'услуг...
2600    ['добрый', 'день', 'оформить', 'кредит', 'банк...
2601    ['самый', 'неудобный', 'интернет', 'банк', 'ну...
2602    ['решить', 'взять', 'втб24', 'кредит', '18', '...
Name: cleaned_review, Length: 2194, dtype: object

### 2. Разделите отзывы на положительные и отрицательные можно на основании значения в поле «score» (например, 1,2- негативные, 4,5-позитивные).

In [ ]:
# Разделение отзывов на положительные и отрицательные
positive_reviews = df[df['score'].isin(['4', '5'])]
positive_reviews['Class'] = 1

negative_reviews = df[df['score'].isin(['1', '2'])]
negative_reviews['Class'] = 0
df_final = pd.concat([positive_reviews, negative_reviews], ignore_index = True)

In [ ]:
df_final

Unnamed: 0                      bank_href             login  \
0           7560  /services/responses/bank/vtb/      Chernik22rus   
1           7562  /services/responses/bank/vtb/             TIMEM   
2           7566  /services/responses/bank/vtb/       arslanovram   
3           7571  /services/responses/bank/vtb/        sakovich25   
4           7572  /services/responses/bank/vtb/  v*******@mail.ru   
...          ...                            ...               ...   
1887       16850                     /bank/vtb/            Юрий,    
1888       16851                     /bank/vtb/           Ольга,    
1889       16852                     /bank/vtb/  Старшова Елена,    
1890       16853                     /bank/vtb/         Наталья,    
1891       16854                     /bank/vtb/   Вернер Сергей,    

                                                 review bank_name  \
0     \n\t\tОбращался в данное отделение за оформлен...       втб   
1     \n\t\tВыражаю свою благодарность Надежде Паньк...       втб   
2     \n\t\tЗдравствуйте, 13 января 2019 года при вн...       втб   
3     \n\t\tбыстрое обслуживание доступные разъяснен...       втб   
4     \n\t\tКогда я пришла оформлять кредит,я попала...       втб   
...                                                 ...       ...   
1887  Санкт–Петербург. Отделение №15. ул. Бутлерова ...       втб   
1888  неправомерное подключение доп.услуг06.08.2016 ...       втб   
1889  Добрый день! Оформлен кредит в Банке ВТБ24 28....       втб   
1890  Самый неудобный интернет банк. Нужно прийти в ...       втб   
1891  Решил взять в ВТБ24 кредит под 18.5%, на 3 год...       втб   

                       site              date                  message_href  \
0           http://banki.ru  22.01.2019 13:45                           NaN   
1           http://banki.ru  12.02.2019 11:40                           NaN   
2           http://banki.ru   19.01.2019 9:55                           NaN   
3           http://banki.ru  22.01.2019 20:06                           NaN   
4           http://banki.ru  05.02.2019 21:58                           NaN   
...                     ...               ...                           ...   
1887  https://www.sravni.ru      19 окт, 2015  /bank/vtb/otzyv/178721/?page   
1888  https://www.sravni.ru      06 дек, 2016  /bank/vtb/otzyv/219166/?page   
1889  https://www.sravni.ru      23 авг, 2015  /bank/vtb/otzyv/170635/?page   
1890  https://www.sravni.ru      23 авг, 2015  /bank/vtb/otzyv/170659/?page   
1891  https://www.sravni.ru      23 авг, 2015  /bank/vtb/otzyv/170668/?page   

     score                                             status  \
0        5                                        Проверяется   
1        5                                        Проверяется   
2        4  =message_href =/services/responses/bank/respon...   
3        5                                        Проверяется   
4        5                                        Проверяется   
...    ...                                                ...   
1887     1                                                NaN   
1888     1                                                NaN   
1889     1                                                NaN   
1890     2                                                NaN   
1891     1                                                NaN   

                  user_href  \
0     /profile/?UID=1579309   
1     /profile/?UID=1686580   
2     /profile/?UID=1565237   
3     /profile/?UID=1582073   
4     /profile/?UID=1652572   
...                     ...   
1887                    NaN   
1888                    NaN   
1889                    NaN   
1890                    NaN   
1891                    NaN   

                                         cleaned_review  Class  
0     ['обращаться', 'данный', 'отделение', 'оформле...      1  
1     ['выражать', 'благодарность', 'надежда', 'пань...      1  
2     ['здравствуйте', '13', 'январь', '2019', 'го

### 3. Постройте векторные представления отзывов с использованием подходов BoW и TF-IDF c разными параметрами (использование n-gramm, различные пороги по отсечению частотных слов), а также с использованием предобучененных векторов word2vec.

In [ ]:
X = df_final['cleaned_review']
y = df_final['Class']

In [ ]:
y = df_final["Class"].tolist()
X = df_final["cleaned_review"].tolist()

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y,  test_size=0.2, random_state=42)


TF-IDF

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Создание объекта TfidfVectorizer с использованием биграмм
tfidf_vectorizer = TfidfVectorizer(ngram_range=(1, 2))

#Применение TF-IDF к текстовым данным тренировочной выборки
tfidf_train_vectors = tfidf_vectorizer.fit_transform(X_train)

#Применение TF-IDF к текстовым данным тестовой выборки
tfidf_test_vectors = tfidf_vectorizer.transform(X_test)

word2vec

In [ ]:
!pip install gensim

In [ ]:
# Train the word2vec model
import gensim
from gensim.models import Word2Vec
w2v_model = gensim.models.Word2Vec(X_train,
                                   vector_size=100,
                                   window=5,
                                   min_count=2)

In [ ]:
import numpy as np

w2v_model.wv.index_to_key
words = set(w2v_model.wv.index_to_key)
X_train_vect = [np.array([w2v_model.wv[i] for i in ls if i in words]) for ls in X_train]
X_test_vect = [np.array([w2v_model.wv[i] for i in ls if i in words]) for ls in X_test]

In [ ]:
# Compute sentence vectors by averaging the word vectors for the words contained in the sentence
X_train_vect_avg = []
for v in X_train_vect:
    if v.size:
        X_train_vect_avg.append(v.mean(axis=0))
    else:
        X_train_vect_avg.append(np.zeros(100, dtype=float))

X_test_vect_avg = []
for v in X_test_vect:
    if v.size:
        X_test_vect_avg.append(v.mean(axis=0))
    else:
        X_test_vect_avg.append(np.zeros(100, dtype=float))

In [ ]:
X_test_vect

[array([[-1.2942616 ,  0.20832627, -0.23475212, ..., -0.05963304,
         -0.06605644,  0.16435444],
        [-0.02544221, -0.12497955,  0.16655833, ...,  0.07564351,
         -0.39607096, -0.53131527],
        [-0.12052625, -0.11971989,  0.6703088 , ...,  0.1979371 ,
          0.26687205, -0.3022085 ],
        ...,
        [-0.4112183 , -0.48197633,  0.21206842, ..., -0.04841724,
          0.17327665,  0.11849228],
        [-0.02544221, -0.12497955,  0.16655833, ...,  0.07564351,
         -0.39607096, -0.53131527],
        [-0.48621604,  0.18996046,  0.13793607, ..., -0.11447394,
         -0.16019328, -0.2562492 ]], dtype=float32),
 array([[-1.2942616 ,  0.20832627, -0.23475212, ..., -0.05963304,
         -0.06605644,  0.16435444],
        [-0.02544221, -0.12497955,  0.16655833, ...,  0.07564351,
         -0.39607096, -0.53131527],
        [ 0.18888132,  0.18933572,  1.1960752 , ...,  0.60402274,
          0.08056459,  0.027778  ],
        ...,
        [ 0.26482174, -0.44446635,  0.4

BoW

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_bow = CountVectorizer(max_features=100, ngram_range=(1, 1))
vectorizer_bow.fit(X_train)

X_train_bow = vectorizer_bow.transform(X_train)
X_test_bow = vectorizer_bow.transform(X_test)

### 4. Классифицируйте отзывы на положительные/отрицательные с использованием разных векторных представлений. Можно использовать любую модель/модели класссификации.

TF-IDF

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()
rf_model = rf.fit(tfidf_train_vectors, y_train)

# Use the trained model to make predictions on the test data
y_pred_tf_idf = rf_model.predict(tfidf_test_vectors)

WORD2VEC

In [ ]:
# Instantiate and fit a basic Random Forest model on top of the vectors

rf = RandomForestClassifier()
#rf_model = rf.fit(X_train_vect_avg, y_train.values.ravel())
rf_model = rf.fit(X_train_vect_avg, y_train)

# Use the trained model to make predictions on the test data
y_pred_w2v = rf_model.predict(X_test_vect_avg)


BoW

In [ ]:
rf = RandomForestClassifier()
rf_model = rf.fit(X_train_bow, y_train)

# Use the trained model to make predictions on the test data
y_pred_bow = rf_model.predict(X_test_bow)

### 5. Проанализируйте точность классификации при использовании разных векторных представлений.

TF-IDF

In [ ]:
from sklearn. metrics import classification_report
print(classification_report(y_test, y_pred_tf_idf))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96       282
           1       1.00      0.76      0.87        97

    accuracy                           0.94       379
   macro avg       0.96      0.88      0.91       379
weighted avg       0.94      0.94      0.94       379



word2vec

In [ ]:
from sklearn. metrics import classification_report
print(classification_report(y_test, y_pred_w2v))

              precision    recall  f1-score   support

           0       0.86      0.95      0.90       282
           1       0.79      0.56      0.65        97

    accuracy                           0.85       379
   macro avg       0.83      0.75      0.78       379
weighted avg       0.84      0.85      0.84       379



BoW

In [ ]:
from sklearn. metrics import classification_report
print(classification_report(y_test, y_pred_bow))

              precision    recall  f1-score   support

           0       0.88      0.98      0.93       282
           1       0.92      0.63      0.75        97

    accuracy                           0.89       379
   macro avg       0.90      0.81      0.84       379
weighted avg       0.90      0.89      0.88       379



Все три модели использовали один и тот же метод классификации - случайный лес, для чистоты эксперимента.

Наилучший показатель точности дала модель, в которой в качестве векторного представления отзывов использоваллся tf-idf подход. Точность данной модели составила 0.94, в то время как BoW дал 0.89, f word2vec - 0.85.